---
title: 손상 된 상태 예외에 대 한 코드 분석 규칙 CA2153 (코드 분석)
description: 손상 된 상태 예외에 대 한 코드 분석 규칙 코드 분석 규칙 CA2153에 대 한 자세한 정보
ms.date: 02/19/2019
ms.topic: reference
author: gewarren
ms.author: gewarren
ms.openlocfilehash: a197e7508c79ba22bd42db6ca36d1fb29fb0c5d6
ms.sourcegitcommit: 2e4adc490c1d2a705a0592b295d606b10b9f51f1
ms.translationtype: MT
ms.contentlocale: ko-KR
ms.lasthandoff: 09/25/2020
ms.locfileid: "96593008"
---
# <a name="ca2153-avoid-handling-corrupted-state-exceptions"></a><span data-ttu-id="db48f-103">CA2153: 손상된 상태 예외 처리 방지</span><span class="sxs-lookup"><span data-stu-id="db48f-103">CA2153: Avoid handling Corrupted State Exceptions</span></span>

| | <span data-ttu-id="db48f-104">값</span><span class="sxs-lookup"><span data-stu-id="db48f-104">Value</span></span> |
|-|-|
| <span data-ttu-id="db48f-105">**규칙 ID**</span><span class="sxs-lookup"><span data-stu-id="db48f-105">**Rule ID**</span></span> |<span data-ttu-id="db48f-106">CA2153</span><span class="sxs-lookup"><span data-stu-id="db48f-106">CA2153</span></span>|
| <span data-ttu-id="db48f-107">**범주**</span><span class="sxs-lookup"><span data-stu-id="db48f-107">**Category**</span></span> |<span data-ttu-id="db48f-108">Microsoft.Security</span><span class="sxs-lookup"><span data-stu-id="db48f-108">Microsoft.Security</span></span>|
| <span data-ttu-id="db48f-109">**수정이 중단 되거나 중단 되지 않습니다.**</span><span class="sxs-lookup"><span data-stu-id="db48f-109">**Fix is breaking or non-breaking**</span></span> |<span data-ttu-id="db48f-110">최신이 아님</span><span class="sxs-lookup"><span data-stu-id="db48f-110">Non-breaking</span></span>|

## <a name="cause"></a><span data-ttu-id="db48f-111">원인</span><span class="sxs-lookup"><span data-stu-id="db48f-111">Cause</span></span>

<span data-ttu-id="db48f-112">[CSEs (손상 된 상태 예외)](/archive/msdn-magazine/2009/february/clr-inside-out-handling-corrupted-state-exceptions) 는 프로세스에 메모리 손상이 있는지를 표시 합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-112">[Corrupted State Exceptions (CSEs)](/archive/msdn-magazine/2009/february/clr-inside-out-handling-corrupted-state-exceptions) indicate that memory corruption exists in your process.</span></span> <span data-ttu-id="db48f-113">프로세스 충돌을 허용하는 대신 catch하면 공격자가 손상된 메모리 영역에 익스플로잇을 배치할 수 있는 경우 보안 취약점이 발생할 수 있습니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-113">Catching these rather than allowing the process to crash can lead to security vulnerabilities if an attacker can place an exploit into the corrupted memory region.</span></span>

## <a name="rule-description"></a><span data-ttu-id="db48f-114">규칙 설명</span><span class="sxs-lookup"><span data-stu-id="db48f-114">Rule description</span></span>

<span data-ttu-id="db48f-115">CSE는 프로세스의 상태가 손상되었으며 시스템에 의해 catch되지 않았음을 나타냅니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-115">CSE indicates that the state of a process has been corrupted and not caught by the system.</span></span> <span data-ttu-id="db48f-116">손상 된 상태 시나리오에서 일반 처리기는 메서드를 특성으로 표시 하는 경우에만 예외를 catch <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute?displayProperty=fullName> 합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-116">In the corrupted state scenario, a general handler only catches the exception if you mark your method with the <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute?displayProperty=fullName> attribute.</span></span> <span data-ttu-id="db48f-117">기본적으로 [CLR (공용 언어 런타임)](../../../standard/clr.md) 은 cses에 대 한 catch 처리기를 호출 하지 않습니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-117">By default, the [Common Language Runtime (CLR)](../../../standard/clr.md) does not invoke catch handlers for CSEs.</span></span>

<span data-ttu-id="db48f-118">가장 안전한 옵션은 이러한 종류의 예외를 catch 하지 않고 프로세스 충돌을 허용 하는 것입니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-118">The safest option is to allow the process to crash without catching these kinds of exceptions.</span></span> <span data-ttu-id="db48f-119">로깅 코드를 사용 해도 공격자가 메모리 손상 버그를 악용할 수 있습니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-119">Even logging code can allow attackers to exploit memory corruption bugs.</span></span>

<span data-ttu-id="db48f-120">이 경고는 모든 예외를 catch 하는 일반 처리기 (예: `catch (System.Exception e)` 또는 `catch` 예외 매개 변수 없음)를 사용 하 여 cses를 catch 할 때 트리거됩니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-120">This warning triggers when catching CSEs with a general handler that catches all exceptions, for example, `catch (System.Exception e)` or `catch` with no exception parameter.</span></span>

## <a name="how-to-fix-violations"></a><span data-ttu-id="db48f-121">위반 문제를 해결하는 방법</span><span class="sxs-lookup"><span data-stu-id="db48f-121">How to fix violations</span></span>

<span data-ttu-id="db48f-122">이 경고를 해결 하려면 다음 중 하나를 수행 합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-122">To resolve this warning, do one of the following:</span></span>

- <span data-ttu-id="db48f-123"><xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute> 특성을 제거합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-123">Remove the <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute> attribute.</span></span> <span data-ttu-id="db48f-124">이는 CSEs가 catch 처리기에 전달 되지 않는 기본 런타임 동작으로 되돌아갑니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-124">This reverts to the default run-time behavior where CSEs are not passed to catch handlers.</span></span>

- <span data-ttu-id="db48f-125">특정 예외 유형을 catch하는 처리기 기본 설정에서 일반 catch 처리기를 제거합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-125">Remove the general catch handler in preference of handlers that catch specific exception types.</span></span> <span data-ttu-id="db48f-126">여기에는 처리기 코드가 안전 하 게 처리할 수 있다고 가정 하는 CSEs가 포함 될 수 있습니다 (드문 경우).</span><span class="sxs-lookup"><span data-stu-id="db48f-126">This may include CSEs, assuming the handler code can safely handle them (rare).</span></span>

- <span data-ttu-id="db48f-127">예외를 호출자에 게 전달 하 고 실행 중인 프로세스를 종료 해야 하는 catch 처리기에서 CSE를 다시 throw 합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-127">Rethrow the CSE in the catch handler, which passes the exception to the caller and should result in ending the running process.</span></span>

## <a name="when-to-suppress-warnings"></a><span data-ttu-id="db48f-128">경고를 표시 하지 않는 경우</span><span class="sxs-lookup"><span data-stu-id="db48f-128">When to suppress warnings</span></span>

<span data-ttu-id="db48f-129">이 규칙에서는 경고를 표시해야 합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-129">Do not suppress a warning from this rule.</span></span>

## <a name="pseudo-code-example"></a><span data-ttu-id="db48f-130">의사 코드 예제</span><span class="sxs-lookup"><span data-stu-id="db48f-130">Pseudo-code example</span></span>

### <a name="violation"></a><span data-ttu-id="db48f-131">위반</span><span class="sxs-lookup"><span data-stu-id="db48f-131">Violation</span></span>

<span data-ttu-id="db48f-132">다음 의사 코드에서는 이 규칙에 의해 검색되는 패턴을 보여 줍니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-132">The following pseudo-code illustrates the pattern detected by this rule.</span></span>

```csharp
[HandleProcessCorruptedStateExceptions]
// Method that handles CSE exceptions.
void TestMethod1()
{
    try
    {
        FileStream fileStream = new FileStream("name", FileMode.Create);
    }
    catch (Exception e)
    {
        // Handle exception.
    }
}
```

### <a name="solution-1---remove-the-attribute"></a><span data-ttu-id="db48f-133">해결 방법 1-특성을 제거 합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-133">Solution 1 - remove the attribute</span></span>

<span data-ttu-id="db48f-134">특성을 제거 <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute> 하면 메서드에서 손상 된 상태 예외를 처리 하지 않습니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-134">Removing the <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute> attribute ensures that Corrupted State Exceptions are not handled by your method.</span></span>

```csharp
void TestMethod1()
{
    try
    {
        FileStream fileStream = new FileStream("name", FileMode.Create);
    }
    catch (Exception e)
    {
        // Handle exception.
    }
}
```

### <a name="solution-2---catch-specific-exceptions"></a><span data-ttu-id="db48f-135">해결 방법 2-특정 예외 catch</span><span class="sxs-lookup"><span data-stu-id="db48f-135">Solution 2 - catch specific exceptions</span></span>

<span data-ttu-id="db48f-136">일반 catch 처리기를 제거하고 특정 예외 형식만 catch합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-136">Remove the general catch handler and catch only specific exception types.</span></span>

```csharp
void TestMethod1()
{
    try
    {
        FileStream fileStream = new FileStream("name", FileMode.Create);
    }
    catch (IOException e)
    {
        // Handle IOException.
    }
    catch (UnauthorizedAccessException e)
    {
        // Handle UnauthorizedAccessException.
    }
}
```

### <a name="solution-3---rethrow"></a><span data-ttu-id="db48f-137">해결 방법 3-다시 throw</span><span class="sxs-lookup"><span data-stu-id="db48f-137">Solution 3 - rethrow</span></span>

<span data-ttu-id="db48f-138">예외를 다시 throw 합니다.</span><span class="sxs-lookup"><span data-stu-id="db48f-138">Rethrow the exception.</span></span>

```csharp
[HandleProcessCorruptedStateExceptions]
void TestMethod1()
{
    try
    {
        FileStream fileStream = new FileStream("name", FileMode.Create);
    }
    catch (Exception e)
    {
        // Rethrow the exception.
        throw;
    }
}
```
